### PR TITLE
Add IBL bad channel detection

### PR DESCRIPTION
**Pull request by JoeZiminski**
PR to port the bad channel detection algorithm of the IBL.

Currently this is not ready due to a number of points I am unsure of below, it makes sense to work through the other IBL PRs first as there are some common issues that will be resolved then. However I'll post this now so its out there.

Some things:

- this assumes the data are in uV (I think this is always true)
- this requires the channels to be in the real (i.e. true spatial) order. I realise now this is not the case, similar to #1059, and will update code ASAP once 1059 is done
- The IBL algorithm [makes the assumption larger-number channels are towards the brain surface](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L236), would this be true for all probes?
- Here [random data is taken](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L119) for all segments to calculate bad channels. Can it be assumed a bad channel will be the same across all segments?
- The IBL function scales [to 1e6 when taking the PSD,](https://github.com/int-brain-lab/ibl-neuropixel/blob/15f85d7d3ec47422f1b5c64bdd47d9d6b1459a09/src/neurodsp/voltage.py#L608) but this is not the case for spikeinterface as units are different. However, to test against IBL it is necessary to scale the PSD only in the function, so there is a [messy workaround](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L180) in the PR where a tag is passed in the random_chunk_kwargs. I think after some re-writing for other reasons, a clearer solution will be apparant.
- [IBL uses N=11 on the median filter](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L224) for channel similarity. Quick testing show this seems to perform well even for 32 or 64 channels, but it is probally worth writing more tests to check this as N=11 seems quite large for 32 channel probe.
- IBL uses NP and [differentiates the psd_hf_threshold ](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L113)by ap vs lf band on the sampling frequency. I guess this is not appropriate for other probes, not sure the best way to handle this.
- the get_random_data_chunks() defaults need to be overwritten, but also would be nice to allow user to edit. There is a [kind of ugly solution](https://github.com/JoeZiminski/spikeinterface/blob/c4f4156a49fa2502b57324c301c5ce928243d8c1/spikeinterface/preprocessing/remove_bad_channels.py#L149), but maybe there is a better way to handle

### CODE REVIEW

**Comment by JoeZiminski** [comment]
In the most recent commits, I swapped the dimensions in the IBL function to avoid the transpose operation. Also, I swapped out the code to calculate the correlation between each channels signal and the reference. The [IBL version](https://github.com/int-brain-lab/ibl-neuropixel/blob/a0e86f1f82610e202564d505307d76eb05c1888a/src/neurodsp/voltage.py#L581) leverages existing code that calculates the cross-correlation and takes the zero-lag output. As we don't otherwise need the code for cross-correlation [I removed most of it](https://github.com/JoeZiminski/spikeinterface/blob/6dd0693b70cbfddaa74dfc30b82ef82a955c4c67/spikeinterface/preprocessing/remove_bad_channels.py#L272).

**Comment by alejoe91** [comment]
@JoeZiminski here is a summary of my changes:

- unified the `detect_bad_channels()` function with two methods:
    - `std`: threshold based on channel std deviations
    - `coherence+psd` (default - IBL method)
- unified tests into `test_remove_bad_channels()`
-  highpass filter if recording is not filtered (see [here](https://github.com/JoeZiminski/spikeinterface/blob/add_ibl_bad_channel_detection/spikeinterface/preprocessing/remove_bad_channels.py#L110-L114))
- sort random chunks by depth if not sorted (needed for finding signals outside of the brain) (see [here](https://github.com/JoeZiminski/spikeinterface/blob/add_ibl_bad_channel_detection/spikeinterface/preprocessing/remove_bad_channels.py#L141-L159))
- exposed all parameters and tried to give a [more interpretable name](https://github.com/JoeZiminski/spikeinterface/blob/add_ibl_bad_channel_detection/spikeinterface/preprocessing/remove_bad_channels.py#L49-L67)
- remove [additional highpass filter,](https://github.com/JoeZiminski/spikeinterface/blob/add_ibl_bad_channel_detection/spikeinterface/preprocessing/remove_bad_channels.py#L227-L228) since data are already filtered

Ready to merge on my side @samuelgarcia

**Review by alejoe91 — APPROVED** [review]

File: spikeinterface/preprocessing/remove_bad_channels.py [hunk 1]
```diff
@@ -1,50 +1,280 @@
...
+    random_data = get_random_data_chunks(recording_hp, **random_chunk_kwargs)
 
+    if method == "std":
         stds = np.std(random_data, axis=0)
-        thresh = bad_threshold * np.median(stds)
-        keep_inds, = np.nonzero(stds < thresh)
+        thresh = std_threshold * np.median(stds)
+        bad_channel_labels = stds > thresh
+        bad_channel_ids = recording.channel_ids[bad_channel_labels]
+        bad_channel_labels = bad_channel_labels.astype(int)
```
Comment (samuelgarcia) [inline spikeinterface/preprocessing/remove_bad_channels.py:131]: Why do we cast bool to int ?

File: spikeinterface/preprocessing/remove_bad_channels.py [hunk 2]
```diff
@@ -1,50 +1,280 @@
...
+        if np.array_equal(np.sort(locs_mono), locs_mono):
+            order_f = None
+            order_r = None
+        else:
+            order_f = np.argsort(locs_mono)
+            order_r = np.argsort(order_f)
+
+        # Create empty channel labels and fill with bad-channel detection estimate for each chunk
+        channel_labels = np.zeros((recording.get_num_channels(),
+                                  recording.get_num_segments() * random_chunk_kwargs["num_chunks_per_segment"]),
```
Comment (samuelgarcia) [inline spikeinterface/preprocessing/remove_bad_channels.py:155]:
```suggestion
                                  len(random_data),
```

File: spikeinterface/preprocessing/remove_bad_channels.py [hunk 3]
```diff
@@ -1,50 +1,280 @@
...
+        Threshold on Nyquist frequency to calculate HF noise band, by default 0.8
+
+    Returns
+    -------
+    1d array
+        Channels labels: 0: good,  1: dead low coherence / amplitude, 2: noisy, 3: outside of the brain
+    """
+    _, nc = raw.shape
+    raw = raw - np.mean(raw, axis=0)[np.newaxis, :]
+    fscale, psd = scipy.signal.welch(raw, fs=fs, axis=0)
```
Comment (samuelgarcia) [inline spikeinterface/preprocessing/remove_bad_channels.py:226]: the output  of scipy.signal.welch heavily depend on nperseg.
This is None by default ans so heavily depend depend on the size of the chunk.
In PSD the window size is super super imprtant.
Here it is silently ingnored I would propose to have in the main function something that clearly ask the user the window size in second and propagte this paramters to here (and also to the random chunk stuff).

File: spikeinterface/preprocessing/remove_bad_channels.py [hunk 4]
```diff
@@ -1,50 +1,280 @@
...
+
+    xcor, xcorr_neighbors, xcorr_distant = compute_correlations(raw, n_neighbors)
```
Comment (samuelgarcia) [inline spikeinterface/preprocessing/remove_bad_channels.py:228]: `compute_correlations` is called only once.
`channels_similarity`is also called once.
This make the maths hard to follow.
I would propose to make a linear code without sub function calls because they are call once.
no ?

File: spikeinterface/preprocessing/remove_bad_channels.py (hunk 1 above)
Comment (JoeZiminski) [reply spikeinterface/preprocessing/remove_bad_channels.py:131]: I think for consistency with IBL method, that also has numbered labels (they are 0-3 though, 0: good, 1: dead, 2: noisy, 3: outside of the brain). But maybe it is confusing to have bool output in one case and int in other

Comment (alejoe91) [reply spikeinterface/preprocessing/remove_bad_channels.py:131]: Yep and that is super useful! Because you want to remove label 3 and interpolate the others ;)

File: spikeinterface/preprocessing/remove_bad_channels.py (hunk 3 above)
Comment (JoeZiminski) [reply spikeinterface/preprocessing/remove_bad_channels.py:226]: That's a good point this is nice to have, in particular if users are changing the psd_hf_threshold. I think in this case using None as default should not have a massive effect as PSD is used only to determine mean PSD > 80% Nyquist (by default). Most window sizes should give sufficient frequency resolution to estimate this well, even if very low e.g. N = 100 (that been said, I can't seem to figure out from SciPy docs what it does default to).

File: spikeinterface/preprocessing/remove_bad_channels.py (hunk 2 above)
Comment (alejoe91) [reply spikeinterface/preprocessing/remove_bad_channels.py]: done

File: spikeinterface/preprocessing/remove_bad_channels.py (hunk 4 above)
Comment (alejoe91) [reply spikeinterface/preprocessing/remove_bad_channels.py]: done

File: spikeinterface/preprocessing/remove_bad_channels.py [hunk 5]
```diff
@@ -1,50 +1,280 @@
...
+        random_chunk_kwargs["concatenated"] = True
+    else:
+        random_chunk_kwargs["return_scaled"] = True
+        random_chunk_kwargs["concatenated"] = False
 
-    def __init__(self, recording, bad_threshold=5, **random_chunk_kwargs):
-        random_data = get_random_data_chunks(recording, **random_chunk_kwargs)
+    random_data = get_random_data_chunks(recording_hp, **random_chunk_kwargs)
 
+    if method == "std":
```
[the remark this answers is not in the document]
Comment (alejoe91) [reply spikeinterface/preprocessing/remove_bad_channels.py]: done

File: spikeinterface/preprocessing/remove_bad_channels.py (hunk 3 above)
Comment (samuelgarcia) [reply spikeinterface/preprocessing/remove_bad_channels.py]: The default in scipy was nperseg=256
I added a parameters to handle this now.
